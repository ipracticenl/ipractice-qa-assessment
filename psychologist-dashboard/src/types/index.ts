// Base types
export interface User {
  id: number;
  name: string;
  type: "psychologist" | "client";
}

// Time range base type
export interface TimeRange {
  id: string;
  from: Date;
  to: Date;
}

// Client DTOs
export interface CreateClientDto {
  name: string;
  initialPsychologistIds: number[];
}

export interface ClientSummaryDto {
  id: number;
  name: string;
}

export interface ClientDetailsDto {
  id: number;
  name: string;
  assignedPsychologists: number[];
  bookedAppointments: Appointment[];
}

// Psychologist DTOs
export interface CreatePsychologistDto {
  name: string;
  initialClients: number[];
}

export interface PsychologistSummaryDto {
  id: number;
  name: string;
}

export interface PsychologistDetailsDto {
  id: number;
  name: string;
  assignedClients: number[];
  availableTimeSlots: AvailableTimeSlot[];
  bookedAppointments: BookedAppointment[];
}

// Booking DTOs
export interface CreateBookingDto {
  psychologistId: number;
  availableTimeSlotId: string;
  comment?: string;
}

export interface ClientDto {
  clientId: number;
}

export interface TimeSlotDto {
  from: Date;
  to: Date;
}

// Model types
export interface AvailableTimeSlot extends TimeRange {
  // Inherits id, from, to from TimeRange
}

export interface Appointment extends TimeRange {
  psychologistId: number;
  comment?: string;
}

export interface BookedAppointment extends TimeRange {
  clientId: number;
}

// Legacy types for backward compatibility
export interface Psychologist extends User {
  type: "psychologist";
  clients: string[];
}

export interface Client extends User {
  type: "client";
  psychologist?: string;
}

export interface Availability {
  id: string;
  psychologistId: number;
  startTime: Date;
  endTime: Date;
  isBooked: boolean;
  clientId?: string;
  clientName?: string;
}
