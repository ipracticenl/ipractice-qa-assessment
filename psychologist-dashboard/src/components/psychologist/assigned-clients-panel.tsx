import { useEffect, useRef, type FC } from "react";
import { Card, CardHeader, CardTitle, CardContent } from "../ui/card";
import {
  TableHeader,
  TableRow,
  TableHead,
  TableBody,
  TableCell,
  Table,
} from "../ui/table";
import { ClientSummaryDto } from "@/types";

interface AssignedClientsPanelProps {
  assignedClients: ClientSummaryDto[];
}

export const AssignedClientsPanel: FC<AssignedClientsPanelProps> = ({
  assignedClients,
}) => {
  const containerRef = useRef<HTMLDivElement>(null);
  const limitPx = 140;

  useEffect(() => {
    const container = containerRef.current as HTMLDivElement;
    if (!container) return;

    function handleScroll() {
      const maxScrollTop =
        container.scrollHeight - container.clientHeight - limitPx;
      if (container.scrollTop > maxScrollTop) {
        container.scrollTop = maxScrollTop;
      }
    }

    container.addEventListener("scroll", handleScroll);
    return () => container.removeEventListener("scroll", handleScroll);
  }, [limitPx]);

  return (
    <Card data-testid="assigned-clients-panel">
      <CardHeader>
        <CardTitle className="flex items-center gap-2">
          <span className="w-5 h-5 bg-green-100 rounded-full flex items-center justify-center text-green-600 text-xs font-bold">
            {assignedClients.length}
          </span>
          Assigned Clients
        </CardTitle>
      </CardHeader>
      <CardContent>
        <div ref={containerRef} className="max-h-[300px] overflow-y-auto">
          <Table disableOverflow>
            <TableHeader className="sticky top-0 z-10 bg-background">
              <TableRow>
                <TableHead>Client Name</TableHead>
                <TableHead>Status</TableHead>
              </TableRow>
            </TableHeader>
            <TableBody>
              {assignedClients.length === 0 ? (
                <TableRow>
                  <TableCell
                    colSpan={2}
                    className="text-center py-8 text-gray-500"
                  >
                    No clients assigned yet
                  </TableCell>
                </TableRow>
              ) : (
                assignedClients.map((client) => (
                  <TableRow
                    key={client.id}
                    data-testid={`client-row-${client.id}`}
                  >
                    <TableCell className="font-medium">{client.name}</TableCell>
                    <TableCell>
                      <span
                        className="inline-flex px-2 py-1 rounded-full text-xs font-medium bg-blue-100 text-blue-800"
                        aria-label="Client is active"
                      >
                        Active
                      </span>
                    </TableCell>
                  </TableRow>
                ))
              )}
            </TableBody>
          </Table>
        </div>
      </CardContent>
    </Card>
  );
};
