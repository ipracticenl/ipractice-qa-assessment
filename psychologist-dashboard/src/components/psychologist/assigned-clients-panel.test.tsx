import { render, screen } from "@testing-library/react";
import { describe, it, expect } from "vitest";
import { AssignedClientsPanel } from "./assigned-clients-panel";
import { ClientSummaryDto } from "@/types";

// Mock data
const mockClients: ClientSummaryDto[] = [
  {
    id: 1,
    name: "John Doe",
  },
  {
    id: 2,
    name: "Jane Smith",
  },
  {
    id: 3,
    name: "Bob Johnson",
  },
];

describe("AssignedClientsPanel", () => {
  describe("Snapshot Tests", () => {
    it("should match snapshot when no clients exist", () => {
      // Given: No clients are provided
      const emptyClients: ClientSummaryDto[] = [];

      // When: The component is rendered with no clients
      const { container } = render(
        <AssignedClientsPanel assignedClients={emptyClients} />
      );

      // Then: The component should match the snapshot
      expect(container).toMatchSnapshot();
    });

    it("should match snapshot when clients exist", () => {
      // Given: Clients are provided
      const clients = mockClients;

      // When: The component is rendered with clients
      const { container } = render(
        <AssignedClientsPanel assignedClients={clients} />
      );

      // Then: The component should match the snapshot
      expect(container).toMatchSnapshot();
    });
  });

  describe("Empty State", () => {
    it("should display empty state message when no clients exist", () => {
      // Given: No clients are provided
      const emptyClients: ClientSummaryDto[] = [];

      // When: The component is rendered
      render(<AssignedClientsPanel assignedClients={emptyClients} />);

      // Then: The empty state message should be displayed
      expect(screen.getByText("No clients assigned yet")).toBeInTheDocument();
    });

    it("should show count as 0 when no clients exist", () => {
      // Given: No clients are provided
      const emptyClients: ClientSummaryDto[] = [];

      // When: The component is rendered
      render(<AssignedClientsPanel assignedClients={emptyClients} />);

      // Then: The count should be displayed as 0
      expect(screen.getByText("0")).toBeInTheDocument();
    });
  });

  describe("Clients Display", () => {
    it("should display clients when they exist", () => {
      // Given: Clients are provided
      const clients = mockClients;

      // When: The component is rendered
      render(<AssignedClientsPanel assignedClients={clients} />);

      // Then: Client rows should be displayed
      expect(screen.getByTestId("client-row-1")).toBeInTheDocument();
      expect(screen.getByTestId("client-row-2")).toBeInTheDocument();
      expect(screen.getByTestId("client-row-3")).toBeInTheDocument();
    });

    it("should display client names correctly", () => {
      // Given: Clients are provided
      const clients = mockClients;

      // When: The component is rendered
      render(<AssignedClientsPanel assignedClients={clients} />);

      // Then: Client names should be displayed
      expect(screen.getByText("John Doe")).toBeInTheDocument();
      expect(screen.getByText("Jane Smith")).toBeInTheDocument();
      expect(screen.getByText("Bob Johnson")).toBeInTheDocument();
    });

    it('should display status as "Active" for all clients', () => {
      // Given: Clients are provided
      const clients = mockClients;

      // When: The component is rendered
      render(<AssignedClientsPanel assignedClients={clients} />);

      // Then: All clients should show "Active" status
      const activeStatuses = screen.getAllByText("Active");
      expect(activeStatuses).toHaveLength(3);
    });

    it("should have correct aria-label for status badges", () => {
      // Given: Clients are provided
      const clients = mockClients;

      // When: The component is rendered
      render(<AssignedClientsPanel assignedClients={clients} />);

      // Then: Status badges should have correct aria-label
      const statusBadges = screen.getAllByLabelText("Client is active");
      expect(statusBadges).toHaveLength(3);
    });

    it("should show correct count in title", () => {
      // Given: Clients are provided
      const clients = mockClients;

      // When: The component is rendered
      render(<AssignedClientsPanel assignedClients={clients} />);

      // Then: The count should be displayed correctly
      expect(screen.getByText("3")).toBeInTheDocument();
    });
  });

  describe("Component Structure", () => {
    it("should render the main panel with correct test ID", () => {
      // Given: Clients are provided
      const clients = mockClients;

      // When: The component is rendered
      render(<AssignedClientsPanel assignedClients={clients} />);

      // Then: The main panel should be rendered with correct test ID
      expect(screen.getByTestId("assigned-clients-panel")).toBeInTheDocument();
    });

    it("should render table headers correctly", () => {
      // Given: Clients are provided
      const clients = mockClients;

      // When: The component is rendered
      render(<AssignedClientsPanel assignedClients={clients} />);

      // Then: The table headers should be displayed
      expect(screen.getByText("Client Name")).toBeInTheDocument();
      expect(screen.getByText("Status")).toBeInTheDocument();
    });

    it("should display the correct title", () => {
      // Given: Clients are provided
      const clients = mockClients;

      // When: The component is rendered
      render(<AssignedClientsPanel assignedClients={clients} />);

      // Then: The title should be displayed
      expect(screen.getByText("Assigned Clients")).toBeInTheDocument();
    });

    it("should have correct CSS classes for styling", () => {
      // Given: Clients are provided
      const clients = mockClients;

      // When: The component is rendered
      const { container } = render(
        <AssignedClientsPanel assignedClients={clients} />
      );

      // Then: The count badge should have correct styling classes
      const countBadge = container.querySelector(
        ".bg-green-100.text-green-600"
      );
      expect(countBadge).toBeInTheDocument();
    });
  });

  describe("Edge Cases", () => {
    it("should handle single client correctly", () => {
      // Given: Only one client is provided
      const singleClient = [mockClients[0]];

      // When: The component is rendered
      render(<AssignedClientsPanel assignedClients={singleClient} />);

      // Then: Only one client row should be displayed
      expect(screen.getByTestId("client-row-1")).toBeInTheDocument();
      expect(screen.queryByTestId("client-row-2")).not.toBeInTheDocument();
      expect(screen.getByText("1")).toBeInTheDocument();
    });

    it("should handle large number of clients", () => {
      // Given: Many clients are provided
      const manyClients = Array.from({ length: 10 }, (_, index) => ({
        id: index + 1,
        name: `Client ${index + 1}`,
      }));

      // When: The component is rendered
      render(<AssignedClientsPanel assignedClients={manyClients} />);

      // Then: All clients should be displayed
      expect(screen.getByText("10")).toBeInTheDocument();
      expect(screen.getByText("Client 1")).toBeInTheDocument();
      expect(screen.getByText("Client 10")).toBeInTheDocument();
    });
  });
});
