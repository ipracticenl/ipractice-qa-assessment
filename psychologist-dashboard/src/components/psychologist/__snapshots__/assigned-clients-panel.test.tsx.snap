// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AssignedClientsPanel > Snapshot Tests > should match snapshot when clients exist 1`] = `
<div>
  <div
    class="rounded-lg border bg-card text-card-foreground shadow-sm"
    data-testid="assigned-clients-panel"
  >
    <div
      class="flex flex-col space-y-1.5 p-6"
    >
      <div
        class="text-2xl font-semibold leading-none tracking-tight flex items-center gap-2"
      >
        <span
          class="w-5 h-5 bg-green-100 rounded-full flex items-center justify-center text-green-600 text-xs font-bold"
        >
          3
        </span>
        Assigned Clients
      </div>
    </div>
    <div
      class="p-6 pt-0"
    >
      <div
        class="overflow-x-auto"
      >
        <div
          class="relative w-full overflow-auto"
        >
          <table
            class="w-full caption-bottom text-sm"
          >
            <thead
              class="[&_tr]:border-b"
            >
              <tr
                class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
              >
                <th
                  class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0"
                >
                  Client Name
                </th>
                <th
                  class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0"
                >
                  Status
                </th>
              </tr>
            </thead>
            <tbody
              class="[&_tr:last-child]:border-0"
            >
              <tr
                class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
                data-testid="client-row-1"
              >
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0 font-medium"
                >
                  John Doe
                </td>
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
                >
                  <span
                    aria-label="Client is active"
                    class="inline-flex px-2 py-1 rounded-full text-xs font-medium bg-blue-100 text-blue-800"
                  >
                    Active
                  </span>
                </td>
              </tr>
              <tr
                class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
                data-testid="client-row-2"
              >
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0 font-medium"
                >
                  Jane Smith
                </td>
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
                >
                  <span
                    aria-label="Client is active"
                    class="inline-flex px-2 py-1 rounded-full text-xs font-medium bg-blue-100 text-blue-800"
                  >
                    Active
                  </span>
                </td>
              </tr>
              <tr
                class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
                data-testid="client-row-3"
              >
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0 font-medium"
                >
                  Bob Johnson
                </td>
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
                >
                  <span
                    aria-label="Client is active"
                    class="inline-flex px-2 py-1 rounded-full text-xs font-medium bg-blue-100 text-blue-800"
                  >
                    Active
                  </span>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AssignedClientsPanel > Snapshot Tests > should match snapshot when no clients exist 1`] = `
<div>
  <div
    class="rounded-lg border bg-card text-card-foreground shadow-sm"
    data-testid="assigned-clients-panel"
  >
    <div
      class="flex flex-col space-y-1.5 p-6"
    >
      <div
        class="text-2xl font-semibold leading-none tracking-tight flex items-center gap-2"
      >
        <span
          class="w-5 h-5 bg-green-100 rounded-full flex items-center justify-center text-green-600 text-xs font-bold"
        >
          0
        </span>
        Assigned Clients
      </div>
    </div>
    <div
      class="p-6 pt-0"
    >
      <div
        class="overflow-x-auto"
      >
        <div
          class="relative w-full overflow-auto"
        >
          <table
            class="w-full caption-bottom text-sm"
          >
            <thead
              class="[&_tr]:border-b"
            >
              <tr
                class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
              >
                <th
                  class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0"
                >
                  Client Name
                </th>
                <th
                  class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0"
                >
                  Status
                </th>
              </tr>
            </thead>
            <tbody
              class="[&_tr:last-child]:border-0"
            >
              <tr
                class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
              >
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0 text-center py-8 text-gray-500"
                  colspan="2"
                >
                  No clients assigned yet
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
